sdks/go/examples/native_wordcap/nativepubsubio/native.go

```diff
@@ -0,0 +1,247 @@
+			case <-timeout.C:
+				log.Debugf(context.Background(), "cancelling receive context, scheduling resumption")
+				cFn()
+				return sdf.ResumeProcessingIn(10 * time.Second), nil
+			}
+		}
+	}
+}
+
+// NativeRead reads messages from a PubSub topic in a streaming context, outputting
+// received messages as a PCollection of byte slices. If the provided subscription
+// name exists for the given topic, the DoFn will read from that subscription; otherwise,
+// a new subscription with the given subscription name will be created and read from.
+//
+// This feature is experimental and subject to change, including its behavior and function signature.
+// Please use the cross-language implementation Read() instead.
+func NativeRead(s beam.Scope, project, topic, subscription string) beam.PCollection {
+	s = s.Scope("pubsubio.NativeRead")
+
+	psRead, err := newPubSubRead(context.Background(), project, topic, subscription)
+	if err != nil {
+		panic(err)
+	}
+	return beam.ParDo(s, psRead, beam.Impulse(s))
+}
+
+// pubSubWrite is a structural DoFn representing writes to a given PubSub topic.
+type pubSubWrite struct {
+	ProjectID string
+	Topic     string
+	client    *pubsub.Client
+}
+
+// Setup initializes a PubSub client if one has not been created already
+func (r *pubSubWrite) Setup(ctx context.Context) error {
+	if r.client == nil {
+		client, err := pubsub.NewClient(context.Background(), r.ProjectID)
+		if err != nil {
+			return err
+		}
+		r.client = client
+	}
+	return nil
+}
+
+// ProcessElement takes a []byte element and publishes it to the provided PubSub
+// topic.
+func (w *pubSubWrite) ProcessElement(ctx context.Context, elm []byte) error {
+	top := w.client.Topic(w.Topic)
+
+	psMess := &pubsub.Message{Data: elm}
+	result := top.Publish(ctx, psMess)
+	if _, err := result.Get(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// newPubSubWrite inserts a write to a PubSub topic into the pipeline.
+func newPubSubWrite(ctx context.Context, projectID, topic string) (*pubSubWrite, error) {
+	if topic == "" {
+		return nil, errors.New("please provide a topic to write to")
+	}
+	client, err := pubsub.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	top := client.Topic(topic)
+	if ok, err := top.Exists(ctx); !ok || err != nil {
+		return nil, fmt.Errorf("failed to get topic; exists: %v, error: %v", ok, err)
+	}
+	return &pubSubWrite{ProjectID: projectID, Topic: top.ID()}, nil
+}
+
+// NativeWrite publishes elements from a PCollection of byte slices to a PubSub topic.
+// If the topic does not exist at pipeline construction time, the function will panic.
+//
+// This feature is experimental and subject to change, including its behavior and function signature.
+// Please use the cross-language implementation Write() instead.
+func NativeWrite(s beam.Scope, col beam.PCollection, project, topic string) {
+	s = s.Scope("pubsubio.NativeWrite")
+
+	psWrite, err := newPubSubWrite(context.Background(), project, topic)
```

Comment:
Same here, have a context passed in, and use that.

Reply:
Done